### PR TITLE
remove assumption about characteristic value handle relationship

### LINKED ISSUE #16946: characteristic value handle vs characteristic handle
This is deep in the weeds, but: Various examples and Zephyr code appear to assume that the handle of the Characteristic Value declaration is one more than the handle of the characteristic declaration.  See e.g. 80bc21e03c5e2397ae3f5a12d154c626aa53b45d from #12254.

`bt_gatt_attr_read_chrc` for example [sets it this way](https://github.com/zephyrproject-rtos/zephyr/blob/13b4c17bf76c1e221d2b64ddc7f5b017440d7619/subsys/bluetooth/host/gatt.c#L981), citing V3G3.3.2 that the Characteristic Value declaration is the first Attribute after the characteristic declaration.  This seems fine if this code is implementing Zephyr's handle assignment policy, but I'm not convinced that's what it does in all cases.

V3G2.5.1 specifies there may be gaps in handles between adjacent attributes, even when the specification requires adjacency.  The O'Reilly "Getting Started with Bluetooth Low Energy" book has (p. 61):

> Although it’s often the case, you should never assume that this *[Characteristic Value]* handle will be contiguous (i.e., 0xNNNN+1 ) to the one containing the declaration.

I noticed this because the `central_hr` sample also assume this when it subscribes using a handle [one past the handle of the characteristc](https://github.com/zephyrproject-rtos/zephyr/blob/13b4c17bf76c1e221d2b64ddc7f5b017440d7619/samples/bluetooth/central_hr/src/main.c#L73).

I'm concerned a non-Zephyr device implementing a standard service may not follow the same policy.

Is what Zephyr's doing really guaranteed to be correct?  The characteristic declaration value [includes the actual value handle](https://github.com/zephyrproject-rtos/zephyr/blob/13b4c17bf76c1e221d2b64ddc7f5b017440d7619/subsys/bluetooth/host/gatt.c#L957), but it's ignored in [parse_characteristic](https://github.com/zephyrproject-rtos/zephyr/blob/13b4c17bf76c1e221d2b64ddc7f5b017440d7619/subsys/bluetooth/host/gatt.c#L2192) where the other information in the value is preserved.

I have a patch that I think fixes this, but want a sanity check on my understanding before going further.

### PR DESCRIPTION
**Pull request by pabigot**
Closes #16946 

## Bluetooth: gatt: avoid assumptions about characteristic handles

Although the Characteristic Value descriptor is required to be
immediately after the characteristic descriptor, the specification
allows for gaps in the corresponding Attribute handles.  Use the value
handle from the characteristic descriptor for value reads.

See BLUETOOTH CORE SPECIFICATION Version 5.1 Vol 3, Part G section 2.5.1
(p. 2345), first paragraph.
### Bluetooth: central_hr: use new API to get value handle

The assumption that the value handle is one past the attribute handle
works for Zephyr servers but may not be true for others.  Avoid the
hard-coded system-specific assumption.

### CODE REVIEW

File: subsys/bluetooth/host/gatt.c [hunk 1]
```diff
@@ -961,12 +961,24 @@ struct gatt_chrc {
 	};
 } __packed;
 
+uint16_t bt_gatt_attr_value_handle(const struct bt_gatt_attr *attr)
+{
+	struct bt_gatt_chrc *chrc = attr->user_data;
+	u16_t handle = chrc->value_handle;
+
```
Comment (Vudentz) [inline subsys/bluetooth/host/gatt.c]: We have to make sure the attr given here is really of the same type we expect which you can probably accomplish by checking if the read callback is bt_gatt_attr_read_chrc, also add an assert to check for NULL attribute.

Comment (pabigot) [reply subsys/bluetooth/host/gatt.c]: done

**Review by Vudentz — APPROVED** [review]
LGTM

**Review by jhedberg — APPROVED** [review]

**Comment by jhedberg** [comment]
@Vudentz was this supposed to have a 1.14 label on it? Seems like a bug (or at least IOP) fix to me. If so a backport PR will need to be created manually.

**Pull request merged**